src/Decoder.php

```diff
@@ -10,6 +10,8 @@ interface Decoder
     /**
      * Decodes from JSON, validating the errors
      *
+     * @param non-empty-string $json
```

Comment:
Some of these diffs come from your `4.x` changes: perhaps best to merge that up first, and then include them here?

Reply:
That's the plan